lib/system/file.go

```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
 	return nil
 }
 
+// Recursively copy
+// dst must always be a directory
+// src may be either a dir or a file
+func CopyAll(src, dst string) (err error, fileCount uint) {
+	fileCount = 0
+	err = copyAll(src, dst, &fileCount)
+	return err, fileCount
+}
+
+func copyAll(src, dst string, fileCount *uint) (err error) {
+	src = filepath.Clean(src)
+	dst = filepath.Clean(dst)
+
+	log.Debugf("CopyAll %s %s", src, dst)
+	si, err := os.Stat(src)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if si.Mode().IsRegular() {
+		*fileCount++
+		return CopyFile(src, filepath.Join(dst, filepath.Base(src)))
+	}
+
+	_, err = os.Stat(dst)
+	if err != nil && !os.IsNotExist(err) {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err != nil && os.IsNotExist(err) {
+		err = os.MkdirAll(dst, si.Mode())
+		if err != nil {
+			return trace.ConvertSystemError(err)
+		}
+	}
+
+	entries, err := ioutil.ReadDir(src)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	for _, entry := range entries {
+		srcPath := filepath.Join(src, entry.Name())
+		dstPath := filepath.Join(dst, entry.Name())
+
+		if entry.IsDir() {
+			err = copyAll(srcPath, dstPath, fileCount)
+			if err != nil {
+				return err
+			}
```

Comment:
You need to `continue` here - otherwise the function will try to `CopyFile` a directory.
On a side note - why reinvent the [`filepath.Walk`](https://golang.org/pkg/path/filepath/#Walk)?